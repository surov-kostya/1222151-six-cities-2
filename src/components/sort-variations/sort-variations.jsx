import React, {PureComponent} from 'react';
import {func, arrayOf} from 'prop-types';
import {variantType} from '../../models/index';

class SortVariations extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      isSortMenuShown: false
    };
    this.sortVariations = props.sortVariations;
    this.onSort = props.onSort;

    this._handleDropdownToggle = this._handleDropdownToggle.bind(this);
  }

  render() {
    return (
      <div className="places__sorting" action="#" method="get">
        <span className="places__sorting-caption">Sort by</span>
        <button className="places__sorting-type" tabIndex="0"
          style={{border: `none`, backgroundColor: `transparent`, outline: `none`}}
          onClick={this._handleDropdownToggle}>
          {this.props.activeSortVariant && this.props.activeSortVariant.name}
          <svg className="places__sorting-arrow" width="7" height="4">
            <use xlinkHref="#icon-arrow-select"></use>
          </svg>
        </button>
        {this.state.isSortMenuShown && <ul className="places__options places__options--custom places__options--opened">
          {this.sortVariations.map((variant, i) => (
            <li
              key={variant.id}
              onClick={() => {
                this.onSort(variant.id);
                this._handleDropdownToggle();
              }}
              className={
                variant.id === this.props.activeSortVariant.id
                  ? `places__option places__option--active`
                  : `places__option`
              }
              tabIndex={i}
            >
              {variant.name}
            </li>
          ))}
        </ul>}
      </div>
    );
  }

  _handleDropdownToggle() {
    this.setState((prevState) => ({isSortMenuShown: !prevState.isSortMenuShown}));
  }
}

SortVariations.propTypes = {
  activeSortVariant: variantType,
  onSort: func,
  sortVariations: arrayOf(variantType)
};

export default SortVariations;
