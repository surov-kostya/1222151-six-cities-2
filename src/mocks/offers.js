import {shape, number, string, arrayOf} from 'prop-types';

export const placeType = shape({
  id: number,
  name: string,
  imageSrc: string,
  contacts: shape({eMail: string}),
  gallerySrcs: arrayOf(string),
  mark: string,
  rating: number,
  parametres: shape({
    entire: string,
    bedrooms: number,
    adults: number
  }),
  price: number,
  features: arrayOf(string),
  host: shape({
    id: number,
    name: string,
    status: string
  }),
  hostsReview: arrayOf(string)
});

export const places = [
  {
    id: 0,
    name: `Beautiful && luxurious apartment at great location`,
    imageSrc: `img/apartment-01.jpg`,
    contacts: {
      eMail: `Oliver.conner@gmail.com`
    },
    gallerySrcs: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/map.jpg`],
    mark: `Premium`,
    rating: 4.8,
    parametres: {
      entire: `Entire place`,
      bedrooms: 3,
      adults: 4
    },
    price: 120,
    features: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
    host: {
      id: 100,
      name: `Angelina`,
      status: `Pro`
    },
    hostsReview: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`
    ]
  },
  {
    id: 1,
    name: `Wood and stone place`,
    imageSrc: `img/apartment-02.jpg`,
    contacts: {
      eMail: `Oliver.conner@gmail.com`
    },
    gallerySrcs: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
    mark: `Premium`,
    rating: 4.8,
    parametres: {
      entire: `Entire place`,
      bedrooms: 3,
      adults: 4
    },
    price: 120,
    features: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
    host: {
      id: 100,
      name: `Angelina`,
      status: `Pro`
    },
    hostsReview: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`
    ]
  },
  {
    id: 2,
    name: `Nice, cozy, warm big bed apartment`,
    imageSrc: `img/apartment-03.jpg`,
    contacts: {
      eMail: `Oliver.conner@gmail.com`
    },
    gallerySrcs: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
    mark: `Premium`,
    rating: 4.8,
    parametres: {
      entire: `Entire place`,
      bedrooms: 3,
      adults: 4
    },
    price: 120,
    features: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
    host: {
      id: 100,
      name: `Angelina`,
      status: `Pro`
    },
    hostsReview: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`
    ]
  },
  {
    id: 3,
    name: `Canal View Prinsengracht`,
    imageSrc: `img/room.jpg`,
    contacts: {
      eMail: `Oliver.conner@gmail.com`
    },
    gallerySrcs: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
    mark: `Premium`,
    rating: 4.8,
    parametres: {
      entire: `Entire place`,
      bedrooms: 3,
      adults: 4
    },
    price: 120,
    features: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
    host: {
      id: 100,
      name: `Angelina`,
      status: `Pro`
    },
    hostsReview: [
      `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`
    ]
  }
];
