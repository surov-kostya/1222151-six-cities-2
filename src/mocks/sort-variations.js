export const variations = [
  {
    id: 0,
    name: `Popular`
  },
  {
    id: 1,
    name: `Price: low to high`
  },
  {
    id: 2,
    name: `Price: high to low`
  },
  {
    id: 3,
    name: `Top rated first`
  }
];
