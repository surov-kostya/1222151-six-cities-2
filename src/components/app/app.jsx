import React from 'react';
import Main from '../main/main';
import PlaceDetails from '../place-details/place-details';
import {arrayOf, func} from 'prop-types';
import {placeType, cityType} from '../../mocks/offers';
import {connect} from 'react-redux';
import {ActionCreator, Operation} from '../../reducer';
import {variantType} from '../../mocks/sort-variations';

const sortPlaces = (places, sortVariant) => {
  switch (sortVariant && sortVariant.name) {
    case `Price: low to high`:
      return places.slice().sort((placeA, placeB) => placeA.price - placeB.price);

    case `Price: high to low`:
      return places.slice().sort((placeA, placeB) => placeB.price - placeA.price);

    case `Top rated first`:
      return places.slice().sort((placeA, placeB) => placeB.rating - placeA.rating);
    default:
      return places;
  }
};

const getPageScreen = ({
  places,
  cities,
  city,
  chooseCity,
  fetchOfferList,
  mainSortVariant,
  sortMain,
  sortVariations
}) => {
  switch (location.pathname) {
    case `/`:
      return <Main
        places={sortPlaces(places, mainSortVariant)}
        cities={cities}
        currentCity={city}
        onTitleClick={() => {}}
        onChooseCity={(chosenCity) => {
          chooseCity(chosenCity);
          fetchOfferList(chosenCity.name);
        }}
        activeSortVariant={mainSortVariant}
        onSort={(variantId) => sortMain(variantId)}
        sortVariations={sortVariations}
      />;
    case `/place-details`:
      return (
        <PlaceDetails
          currentCity={city}
          place={places[0]}
          neighbors={
            places[0].neighborIds
              .map((id) => places.find((place) => place.id === id))
          }
        />
      );
    default:
      return (
        <h1>Page not found</h1>
      );
  }
};

getPageScreen.propTypes = {
  places: arrayOf(placeType),
  cities: arrayOf(cityType),
  city: cityType,
  chooseCity: func,
  fetchOfferList: func,
  mainSortVariant: variantType,
  sortMain: func,
  sortVariations: arrayOf(variantType)
};

const App = (props) => {
  return getPageScreen(props);
};

App.propTypes = {
  places: arrayOf(placeType),
  cities: arrayOf(cityType),
  city: cityType,
  chooseCity: func,
  fetchOfferList: func,
  mainSortVariant: variantType,
  sortMain: func,
  sortVariations: arrayOf(variantType)
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, state);
const mapDispatchToProps = (dispatch) => ({
  chooseCity: (city) => dispatch(ActionCreator.chooseCity(city)),
  fetchOfferList: (cityName) => dispatch(Operation.fetchOfferList(cityName)),
  sortMain: (variantId) => dispatch(ActionCreator.sortMain(variantId))
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
