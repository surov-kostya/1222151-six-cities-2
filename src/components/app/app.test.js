import React from 'react';
import renderer from 'react-test-renderer';
import {App} from './app';
import {cities} from '../../mocks/offers';
import {variations} from '../../mocks/sort-variations';

it(`App renders correctly`, () => {
  const tree = renderer
    .create(<App
      places={cities[0].places}
      cities={cities.map((city) => ({id: city.id, name: city.name, coords: city.coords}))}
      city={{id: 0, name: `Paris`, coords: [1, 2]}}
      chooseCity={() => {}}
      fetchOfferList={() => {}}
      sortVariations={variations}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
