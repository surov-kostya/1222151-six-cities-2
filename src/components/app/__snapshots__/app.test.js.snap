// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
Array [
  "",
  <div
    className="page page--gray page--main"
  >
    <header
      className="header"
    >
      <div
        className="container"
      >
        <div
          className="header__wrapper"
        >
          <div
            className="header__left"
          >
            <a
              className="header__logo-link header__logo-link--active"
              href="/"
              onClick={[Function]}
            >
              <img
                alt="6 cities logo"
                className="header__logo"
                height="41"
                src="img/logo.svg"
                width="81"
              />
            </a>
          </div>
          <nav
            className="header__nav"
          >
            <ul
              className="header__nav-list"
            >
              <li
                className="header__nav-item user"
              >
                <a
                  className="header__nav-link header__nav-link--profile"
                  href="/favorites"
                  onClick={[Function]}
                >
                  <div
                    className="header__avatar-wrapper user__avatar-wrapper"
                    style={
                      Object {
                        "backgroundImage": "/static/avatar/2.jpg",
                      }
                    }
                  />
                  <span
                    className="header__user-name user__name"
                  >
                    q@q.ru
                  </span>
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
    <main
      className="page__main page__main--index page__main--index-empty"
    >
      <h1
        className="visually-hidden"
      >
        Cities
      </h1>
      <div
        className="tabs"
      >
        <section
          className="locations container"
        >
          <ul
            className="locations__list tabs__list"
          >
            <li
              className="locations__item"
            >
              <a
                className="locations__item-link tabs__item tabs__item--active"
                onClick={[Function]}
              >
                <span>
                  Paris
                </span>
              </a>
            </li>
            <li
              className="locations__item"
            >
              <a
                className="locations__item-link tabs__itemfalse"
                onClick={[Function]}
              >
                <span>
                  Cologne
                </span>
              </a>
            </li>
            <li
              className="locations__item"
            >
              <a
                className="locations__item-link tabs__itemfalse"
                onClick={[Function]}
              >
                <span>
                  Brussels
                </span>
              </a>
            </li>
            <li
              className="locations__item"
            >
              <a
                className="locations__item-link tabs__itemfalse"
                onClick={[Function]}
              >
                <span>
                  Amsterdam
                </span>
              </a>
            </li>
          </ul>
        </section>
      </div>
      <div
        className="cities"
      >
        <div
          className="cities__places-container container"
        >
          <section
            className="cities__places places"
          >
            <h2
              className="visually-hidden"
            >
              Places
            </h2>
            <b
              className="places__found"
            >
              1
               places to stay in 
              Paris
            </b>
            <div
              action="#"
              className="places__sorting"
              method="get"
            >
              <span
                className="places__sorting-caption"
              >
                Sort by
              </span>
              <button
                className="places__sorting-type"
                onClick={[Function]}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "border": "none",
                    "outline": "none",
                  }
                }
                tabIndex="0"
              >
                Popular
                <svg
                  className="places__sorting-arrow"
                  height="4"
                  width="7"
                >
                  <use
                    xlinkHref="#icon-arrow-select"
                  />
                </svg>
              </button>
            </div>
            <div
              className="cities__places-list tabs__content"
            >
              <article
                className="cities__place-card place-card"
                onMouseEnter={[Function]}
              >
                <div
                  className="place-card__mark"
                >
                  <span>
                    Premium
                  </span>
                </div>
                <div
                  className="cities__image-wrapper place-card__image-wrapper"
                >
                  <a
                    href="/offer/Paris/101"
                    onClick={[Function]}
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height="200"
                      src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/6.jpg"
                      width="260"
                    />
                  </a>
                </div>
                <div
                  className="place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        931
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button place-card__bookmark-button--active button"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "48%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      href="/offer/Paris/101"
                      onClick={[Function]}
                    >
                      Loft Studio in the Central Area
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  />
                </div>
              </article>
            </div>
          </section>
          <div
            className="cities__right-section"
          >
            <section
              className="cities__map map"
            >
              <div
                id="map"
                style={
                  Object {
                    "height": "100%",
                  }
                }
              />
            </section>
          </div>
        </div>
      </div>
    </main>
  </div>,
]
`;
